src/main/scala/mesosphere/marathon/core/storage/store/impl/cache/LazyCachingPersistenceStore.scala

```diff
@@ -140,22 +169,61 @@ class LazyCachingPersistenceStore[K, Category, Serialized](
     val category = ir.category
     val storageId = ir.toStorageId(id, None)
     lockManager.executeSequentially(category.toString) {
-      async { // linter:ignore UnnecessaryElseBranch
-        await(store.store(id, v, version))
-        val old = idCache.getOrElse(category, Nil) // linter:ignore UndesirableTypeInference
-        idCache.put(category, id +: old)
-        Done
+      lockManager.executeSequentially(storageId.toString) {
+        async { // linter:ignore UnnecessaryElseBranch
+          await(store.store(id, v, version))
+          updateCachedVersions(storageId, version, category, v)
+          Done
+        }
       }
     }
   }
 
-  override def versions[Id, V](id: Id)(implicit ir: IdResolver[Id, V, Category, K]): Source[OffsetDateTime, NotUsed] =
-    store.versions(id)
+  protected[cache] def maybePurgeCachedVersions(toRemove: Int = 500, pRemoval: Double = 0.05)(f: () => Boolean): Unit =
+    while (f()) {
+      // randomly GC the versions
+      val counter = new AtomicInteger
+      versionedValueCache.retain { (k, v) =>
+        val x = scala.util.Random.nextDouble()
```

Comment:
Still don't like this logic. Its really strange and counter intuitive - and you still didn't address my earlier comment that "f" shouldn't be passed in in the first place. (You can test this stuff by mocking or wrapping InMemoryPersistenceStore). I still think the atomic is 100% unnecessary and can be refactored out differently. If you're really considered about the "remove 1 element" then remove X% of elements as configured. I'm hoping there is a more efficient way to do so.

In addition, this should ultimately require a lock on the entire collection, thus all reads and writes for versioning should go through this lock - and not through the other locks in the non-versioned places.